src/string_impls/from_string.rs

```diff
@@ -0,0 +1,638 @@
+		return Err(ParseComplexError::InvalidStartingCharacterInIdent.at_one(start.offset()));
+	}
+	t.skip_tokens_while(|c| c.is_alphanumeric() || *c == '_');
+	let end = t.location();
+
+	let ident_str = t.slice(start, end).collect();
+	Ok(ident_str)
+}
+
+// Skip any whitespace characters
+fn skip_whitespace(t: &mut impl Tokens<Item = char>) {
+	t.skip_tokens_while(|c| c.is_whitespace());
+}
+
+// Skip a provided separator, with optional spaces on either side
+fn skip_spaced_separator(t: &mut impl Tokens<Item = char>, s: char) -> bool {
+	skip_whitespace(t);
+	let is_sep = t.token(s);
+	skip_whitespace(t);
+	is_sep
+}
+
+// Turn a ` Result<T, Option<E>>` into `Option<Result<T, E>>`.
+fn transpose_err<T, E>(r: Result<T, Option<E>>) -> Option<Result<T, E>> {
+	match r {
+		Ok(val) => Some(Ok(val)),
+		Err(Some(e)) => Some(Err(e)),
+		Err(None) => None,
+	}
+}
+
+#[cfg(test)]
+mod test {
+	use super::*;
+
+	fn from(s: &str) -> Result<Value<()>, ParseError> {
+		let (res, remaining) = from_str(s);
+		if res.is_ok() {
+			// all successful parse tests fully consume the input string:
+			assert_eq!(remaining.len(), 0, "was not expecting any unparsed output");
+		}
+		res
+	}
+
+	#[test]
+	fn parse_bools() {
+		assert_eq!(from("true"), Ok(Value::bool(true)));
+		assert_eq!(from("false"), Ok(Value::bool(false)));
+	}
+
+	#[test]
+	fn parse_numbers() {
+		assert_eq!(from("123"), Ok(Value::uint(123u128)));
+		assert_eq!(from("1_234_56"), Ok(Value::uint(123_456_u128)));
+		assert_eq!(from("+1_234_56"), Ok(Value::uint(123_456_u128)));
+		assert_eq!(from("-123_4"), Ok(Value::int(-1234)));
+		assert_eq!(from("-abc"), Err(ParseNumberError::ExpectedDigit.between(1, 2)));
+	}
+
+	#[test]
+	fn parse_chars() {
+		assert_eq!(from("'a'"), Ok(Value::char('a')));
+		assert_eq!(from("'😀'"), Ok(Value::char('😀')));
+		assert_eq!(from("'\\n'"), Ok(Value::char('\n')));
+		assert_eq!(from("'\\t'"), Ok(Value::char('\t')));
+		assert_eq!(from("'\\\"'"), Ok(Value::char('"')));
+		assert_eq!(from("'\\\''"), Ok(Value::char('\'')));
+		assert_eq!(from("'\\r'"), Ok(Value::char('\r')));
+		assert_eq!(from("'\\\\'"), Ok(Value::char('\\')));
+		assert_eq!(from("'\\0'"), Ok(Value::char('\0')));
+		assert_eq!(from("'a"), Err(ParseCharError::ExpectedClosingQuoteToMatch(0).at_one(2)));
+	}
+
+	#[test]
+	fn parse_strings() {
+		assert_eq!(from("\"\\n \\r \\t \\0 \\\"\""), Ok(Value::string("\n \r \t \0 \"")));
+		assert_eq!(from("\"Hello there 😀\""), Ok(Value::string("Hello there 😀")));
+		assert_eq!(from("\"Hello\\n\\t there\""), Ok(Value::string("Hello\n\t there")));
+		assert_eq!(from("\"Hello\\\\ there\""), Ok(Value::string("Hello\\ there")));
+		assert_eq!(
+			from("\"Hello\\p there\""),
+			Err(ParseStringError::ExpectedValidEscapeCode.between(7, 8))
+		);
+		assert_eq!(from("\"Hi"), Err(ParseStringError::ExpectedClosingQuoteToMatch(0).at_one(3)));
+	}
+
+	#[test]
+	fn parse_unnamed_composites() {
+		assert_eq!(
+			from("(  true, 1234 ,\t\n\t \"Hello!\" )"),
```

Comment:
What do you think about allowing `[ ]` delimiters for unnamed composites so it allows the rust array syntax for sequences? 

I suppose the issue is that it can't be done in the other direction with `to_string` since there is no type information available.

Reply:
Personally I would miss it when entering values for vecs/arrays/slices since it does not fit the intuition of having Rust like representations of values.

However as you say it does require some thinking about to make it consistent and we can do without it for now.

Reply:
The way I approached this is that the type information (array type and length) is all a part of the scale-info stuff, and on the value side all we need is an unnamed composite type which happens to contain the right types and length and it'll happily encode to that thing. It's just an extension of my general policy of not requiring any type information to exist in the runtime values to keep them fairly simple/flexible.

Reply:
So, if you try encoding a `Composite::Unnamed` into eg a u8 array, as long as the length lines up and the values in the `Composite::Unnamed` are numbers that fit inside a `u8`, it'll encode happily (and depending on the target type, either we'll insert that length prefix if it's a sequence or not if it's an array)